instrumentation/vertx-web-3.0/src/main/java/io/opentelemetry/javaagent/instrumentation/hypertrace/vertx/HttpRequestInstrumentation.java

```diff
@@ -0,0 +1,261 @@
+        }
+      }
+    }
+
+    @Advice.OnMethodExit(suppress = Throwable.class, onThrowable = Throwable.class)
+    public static void exit() {
+      CallDepthThreadLocalMap.decrementCallDepth(HttpClientRequest.class);
+    }
+  }
+
+  public static class EndRequestAdvice_buffer {
+    @Advice.OnMethodEnter(suppress = Throwable.class)
+    public static void enter(
+        @Advice.This HttpClientRequest request, @Advice.Argument(0) Buffer chunk)
+        throws IOException {
+
+      int callDepth = CallDepthThreadLocalMap.incrementCallDepth(HttpClientRequest.class);
+      if (callDepth > 0) {
+        return;
+      }
+
+      Contexts contexts =
+          InstrumentationContext.get(HttpClientRequest.class, Contexts.class).get(request);
+      if (contexts == null) {
+        return;
+      }
+      Span span = Span.fromContext(contexts.context);
+
+      String contentType = request.headers().get("Content-Type");
+      if (HypertraceConfig.get().getDataCapture().getHttpBody().getRequest().getValue()
+          && ContentTypeUtils.shouldCapture(contentType)) {
+
+        BoundedCharArrayWriter buffer =
+            InstrumentationContext.get(MultiMap.class, BoundedCharArrayWriter.class)
+                .get(request.headers());
+        if (buffer == null) {
+          span.setAttribute(
+              HypertraceSemanticAttributes.HTTP_REQUEST_BODY,
+              chunk.toString(StandardCharsets.UTF_8.name()));
+        } else {
+          buffer.write(chunk.toString(StandardCharsets.UTF_8.name()));
+          span.setAttribute(HypertraceSemanticAttributes.HTTP_REQUEST_BODY, buffer.toString());
+        }
+      }
+    }
+
+    @Advice.OnMethodExit(suppress = Throwable.class, onThrowable = Throwable.class)
+    public static void exit() {
+      CallDepthThreadLocalMap.decrementCallDepth(HttpClientRequest.class);
+    }
+  }
+
+  public static class WriteRequestAdvice_string {
+    @Advice.OnMethodEnter(suppress = Throwable.class)
+    public static void enter(
+        @Advice.This HttpClientRequest request, @Advice.Argument(0) String chunk)
+        throws IOException {
+
+      int callDepth = CallDepthThreadLocalMap.incrementCallDepth(HttpClientRequest.class);
+      if (callDepth > 0) {
+        return;
+      }
+
+      String contentType = request.headers().get("Content-Type");
+      if (HypertraceConfig.get().getDataCapture().getHttpBody().getRequest().getValue()
+          && ContentTypeUtils.shouldCapture(contentType)) {
+
+        ContextStore<MultiMap, BoundedCharArrayWriter> contextStore =
+            InstrumentationContext.get(MultiMap.class, BoundedCharArrayWriter.class);
+        BoundedCharArrayWriter buffer = contextStore.get(request.headers());
+        if (buffer == null) {
+          buffer = BoundedBuffersFactory.createWriter();
+          contextStore.put(request.headers(), buffer);
+        }
+        buffer.write(chunk);
```

Comment:
Why are we calling this here and in EndRequestAdvice both?

Reply:
yes, this writes a chunk and in the end method it as well writes a chunk and sends the request.